src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs

```diff
@@ -156,10 +157,13 @@ private bool Visit(ModelMetadata metadata, string key, object model)
         {
             RuntimeHelpers.EnsureSufficientExecutionStack();
 
-            if (model != null && !_currentPath.Add(model))
+            if (model != null)
             {
-                // This is a cycle, bail.
-                return true;
+                if (!_currentPath.Push(model))
```

Comment:
Why not leave the condition like before `if (model != null && !_currentPath.Add(model))`?